src/producer.rs

```diff
@@ -1,242 +1,343 @@
+            },
+        }
+    }
+}
 
-                let physical_gcno_path = path.with_file_name(format!("{}_{}.gcno", stem, 1));
+fn archive_producer(tmp_dir: &Path,
+                    gcnos: RefCell<HashMap<String, &Archive>>,
+                    gcdas: RefCell<HashMap<String, Vec<&Archive>>>,
+                    queue: &WorkQueue,
+                    ignore_orphan_gcno: bool,
+                    is_llvm: bool) {
+
+    for (stem, archive) in gcnos.borrow().iter() {
+        match gcdas.borrow().get(stem) {
+            Some(archs) => {
+                let archive = *archive;
+                let gcno = format!("{}.gcno", stem).to_string();
+                let physical_gcno_path = tmp_dir.join(format!("{}_{}.gcno", stem, 1));
                 let mut gcno_buf_opt: Option<Arc<Vec<u8>>> = if is_llvm {
                     let mut buffer: Vec<u8> = Vec::new();
-                    gcno_file.read_to_end(&mut buffer).expect("Failed to read gcno file");
+                    archive.read_in_buffer(&gcno, &mut buffer);
                     Some(Arc::new(buffer))
                 } else {
-                    extract_file(&mut gcno_file, &physical_gcno_path);
+                    archive.extract(&gcno, &physical_gcno_path);
                     None
                 };
 
-                let gcda_path_in_zip = gcno_path_in_zip.with_extension("gcda");
-
-                for (num, &mut (gcda_archive_name, ref mut gcda_archive)) in gcda_archives.iter_mut().enumerate() {
-                    let gcno_path = path.with_file_name(format!("{}_{}.gcno", stem, num + 1));
+                for (num, &gcda_arch) in archs.iter().enumerate() {
+                    let gcno_path = tmp_dir.join(format!("{}_{}.gcno", stem, num + 1));
+                    let gcda = format!("{}.gcda", stem).to_string();
 
-                    if let Ok(mut gcda_file) = gcda_archive.by_name(&gcda_path_in_zip.to_str().unwrap().replace("\\", "/")) {
-                        match gcno_buf_opt {
-                            Some(ref gcno_buf) => {
-                                let mut gcda_buf: Vec<u8> = Vec::new();
-                                gcda_file.read_to_end(&mut gcda_buf).expect("Failed to read gcda file");
-                                let gcno_stem = path.with_file_name(format!("{}_{}", stem, num + 1));
-                                let gcno_stem = gcno_stem.to_str().expect("Failed to create stem file string");
+                    match gcno_buf_opt {
+                        Some(ref gcno_buf) => {
+                            let mut gcda_buf: Vec<u8> = Vec::new();
+                            let gcno_stem = tmp_dir.join(format!("{}_{}", stem, num + 1));
+                            let gcno_stem = gcno_stem.to_str().expect("Failed to create stem file string");
 
+                            if gcda_arch.read_in_buffer(&gcda, &mut gcda_buf) || (num == 0 && !ignore_orphan_gcno) {
                                 queue.push(Some(WorkItem {
```

Comment:
Maybe we can move the queue.push calls outside the match, so that we can share most of it?

Reply:
the two queue.push are preceeded by a "if gcda_archive.read_in_buffer..." and "if gcda_archive.extract(&gcda, &gcda_path)". The two conditions can be false because we don't find any gcda in archive which is ok when orphan are allowed. I could return an Option from the match but that means that we need to check if it's Some or not because pushing a None will stop everything (see lib::consumer). So I'm not sure we'll win something to put the queue.push outside the match.